src/GSoft.Extensions.Http.Authentication.ClientCredentialsGrant/ClientCredentialsTokenCache.cs

```diff
@@ -0,0 +1,73 @@
+﻿using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Options;
+
+namespace GSoft.Extensions.Http.Authentication.ClientCredentialsGrant;
+
+/// <summary>
+/// An implementation of a token cache made of two cache levels.
+/// We use a memory cache (L1) for faster look up, then a distributed cache (L2).
+/// </summary>
+internal sealed class ClientCredentialsTokenCache : IClientCredentialsTokenCache
+{
+    private readonly IMemoryCache _memoryCache;
+    private readonly IDistributedCache _distributedCache;
+    private readonly IClientCredentialsTokenSerializer _tokenSerializer;
+    private readonly IOptionsMonitor<ClientCredentialsOptions> _optionsMonitor;
+
+    public ClientCredentialsTokenCache(IDistributedCache distributedCache, IClientCredentialsTokenSerializer tokenSerializer, IOptionsMonitor<ClientCredentialsOptions> optionsMonitor)
+        : this(new MemoryCache(new MemoryCacheOptions()), distributedCache, tokenSerializer, optionsMonitor)
+    {
+    }
+
+    // Constructor used by tests
+    internal ClientCredentialsTokenCache(IMemoryCache memoryCache, IDistributedCache distributedCache, IClientCredentialsTokenSerializer tokenSerializer, IOptionsMonitor<ClientCredentialsOptions> optionsMonitor)
+    {
+        this._memoryCache = memoryCache;
+        this._distributedCache = distributedCache;
+        this._tokenSerializer = tokenSerializer;
+        this._optionsMonitor = optionsMonitor;
+    }
+
+    public async Task SetAsync(string clientName, ClientCredentialsToken token, CancellationToken cancellationToken)
+    {
+        var options = this._optionsMonitor.Get(clientName);
+
+        var tokenBytes = this._tokenSerializer.Serialize(clientName, token);
+        var absoluteExpiration = token.Expiration.Subtract(options.CacheLifetimeBuffer);
+
+        // Store token in L1 first
+        this._memoryCache.Set(options.CacheKey, tokenBytes, new MemoryCacheEntryOptions { AbsoluteExpiration = absoluteExpiration });
+
+        // Then store token in L2
+        var distributedCacheOptions = new DistributedCacheEntryOptions { AbsoluteExpiration = absoluteExpiration };
+        await this._distributedCache.SetAsync(options.CacheKey, tokenBytes, distributedCacheOptions, cancellationToken).ConfigureAwait(false);
+    }
+
+    public async Task<ClientCredentialsToken?> GetAsync(string clientName, CancellationToken cancellationToken)
+    {
+        var options = this._optionsMonitor.Get(clientName);
```

Comment:
You mean change the cache key to append like "L1" or "L2"?

Microsoft's `IMemoryCache` implementation uses `ConcurrentDictionary` under the hood and this is the fastest way to store data in-memory with thread safety. I don't really see how we could improve this. We're also talking about CPU optimizations in the scale of nanoseconds, there are probably better ways to improve performance overall...

But before talking about performance we first need to demonstrate that there is a performance issue with benchmarks, and also show that it can have an impact on applications at runtime, which I really doubt.

What we have here is a smaller version of MSAL's web token L1L2 cache that can be found [here](https://github.com/AzureAD/microsoft-identity-web/blob/2.9.0/src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs).

Reply:
From what I understood from the code, you first try the L1 and then L2, your object could have an index indicating which level of cache it is stored in to reduce misses. But indeed this is somewhat very small for one miss. If there is a benchmark I would be curious to see a scenario with no misses and a scenario with only misses.

Reply:
I'll close it for now since the gain in performance is indeed very small. For future versions it might be worth reconsidering.